boot.c: clear current tread id during initialization

lorc opened the pull request:
When OP-TEE is built with CFG_VIRTUALIZATION=y, it does not call
`thread_clr_boot_thread()` during boot because the threads are
allocated in "tee" memory area, which is not available when there is
no virtual guests.

So, in this case local core state is left in erroneous state, which
causes assertion violation in thread_alloc_and_run(), when guests
calls OP-TEE for the first time from boot core.

Fixes: b166fabf3e8c ("core: initialize thread_core_local::curr_thread to -1")
Signed-off-by: Volodymyr Babchuk <volodymyr_babchuk@epam.com>

<!--
    If you are new to submitting pull requests to OP-TEE, then please have a
    look at the list below and tick them off before submitting the pull request.

    1. Read our contribution guidelines:
         https://optee.readthedocs.io/en/latest/general/contribute.html

    2. Read the contribution section in Notice.md and pay extra attention to the
       "Developer Certificate of Origin" in the contribution guidelines.

    3. You should run checkpatch preferably before submitting the pull request.

    4. When everything has been reviewed, you will need to squash, rebase and
       add tags like `Reviewed-by`, `Acked-by`, `Tested-by` etc. More details
       about this can also be found on the link provided above.

    NOTE: This comment will not be shown in the pull request, so no harm keeping
    it, but feel free to remove it if you like.
-->

jenswi-linaro commented on core/arch/arm/kernel/boot.c:
```diff
@@ -1152,6 +1152,7 @@ static void init_primary(unsigned long pageable_part, unsigned long nsec_entry)
 	 */
 	thread_get_core_local()->curr_thread = 0;
 	init_runtime(pageable_part);
+	thread_get_core_local()->curr_thread = -1;
```
This is only needed with `CFG_VIRTUALIZATION` and `thread_init_boot_thread()` immediately undoes this.
Perhaps we could do it like:
``` C
#ifdef CFG_VIRTUALIZATION
	/* A nice comment explaining why this is needed */
	thread_get_core_local()->curr_thread = -1;
#else
	thread_init_boot_thread();
#endif

jforissier replied:
`if IS_ENABLED(CFG_VIRTUALIZATION)` would work in this case

jforissier commented:
s/tread/thread/ in the subject and perhaps a `virt:` prefix would be nice.

Sorry for introducing the regression, I have no way of testing with virtualization enabled :-/

jforissier commented:
@lorc I have just merged https://github.com/OP-TEE/optee_os/pull/3998 which changes a few related things, could you please rebase and check if a fix is still needed for virtualization? Thanks.

lorc commented:
> Sorry for introducing the regression, I have no way of testing with virtualization enabled :-/

Yep. I'm looking on how to automate testing. Long time ago I had a fully automated build for QEMU with Xen... I believe, I can redo it again.
I'm just not sure, if Shippable would have resources to run it. IIRC, it has some time limit, right?

> I have just merged #3998 which changes a few related things, could you please rebase and check if a fix is still needed for virtualization? Thanks.

I can confirm that patch is needed. But, apparently, we need some other fix. Because, with my patch applied, boot process just stalls somewhere. I'm looking at this.

lorc commented:
> I'm looking at this.

Oops, that was my mistake in reworked patch. Now it should work fine. I addressed all the comments.

BTW, why `checkpatch` complaints about `IS_ENABLED`?

`WARNING: IS_ENABLED(CFG_VIRTUALIZATION) is normally used as IS_ENABLED(CONFIG_CFG_VIRTUALIZATION)
`

jforissier commented:
> BTW, why `checkpatch` complaints about `IS_ENABLED`?
> 
> `WARNING: IS_ENABLED(CFG_VIRTUALIZATION) is normally used as IS_ENABLED(CONFIG_CFG_VIRTUALIZATION) `

Oh, that's annoying :-/ This new warning was introduced [recently](https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/commit/?id=50161266973bcc662e969e63d68fc7bff71de21b) and 
 cannot be turned off, unfortunately. The Linux kernel uses `CONFIG_` as a prefix for configuration options while OP-TEE uses `CFG_` so it confuses checkpatch.

I think we have to live with it for now. Longer term, perhaps we can propose a patch to make checkpatch aware of the [`CONFIG_`](https://www.kernel.org/doc/html/latest/kbuild/kconfig.html#config) environment variable.

jforissier commented:
`Acked-by: Jerome Forissier <jerome@forissier.org>`

jforissier commented:
Sorry I posted my Ack to quickly!

The correct syntax is `if IS_ENABLED(CFG_VIRTUALIZATION) {` (a "normal" `if`, not a preprocessor thing) with `#include <config.h>`.

lorc commented:
> The correct syntax is `if (IS_ENABLED(CFG_VIRTUALIZATION)) { `

Oh, I see. Should be fine now.

jforissier commented:
Thanks, looks good indeed.
`Acked-by: Jerome Forissier <jerome@forissier.org>`

lorc commented:
Added the tag

jforissier commented:
> Longer term, perhaps we can propose a patch to make checkpatch aware of the [`CONFIG_`](https://www.kernel.org/doc/html/latest/kbuild/kconfig.html#config) environment variable.

https://lkml.org/lkml/2020/8/17/537
